File: 3740-minimum-distance-between-three-equal-elements-i/solution.cpp
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <climits>

using namespace std;

class Solution {
public:
    /**
     * Calculates the minimum distance of a good tuple (i, j, k).
     * The distance simplifies to 2 * (max_index - min_index).
     */
    int minimumDistance(vector<int>& nums) {
        int n = nums.size();
        // Map to store indices for each value encountered
        unordered_map<int, vector<int>> valueToIndices;
        
        for (int i = 0; i < n; ++i) {
            valueToIndices[nums[i]].push_back(i);
        }
        
        int minDistance = INT_MAX;
        bool found = false;
        
        for (auto const& [val, indices] : valueToIndices) {
            // A good tuple requires at least 3 occurrences of the same value
            if (indices.size() >= 3) {
                found = true;
                // To minimize 2 * (indices[k] - indices[i]), we check 
                // consecutive triplets in the sorted indices list.
                for (size_t i = 0; i <= indices.size() - 3; ++i) {
                    int currentDist = 2 * (indices[i + 2] - indices[i]);
                    if (currentDist < minDistance) {
                        minDistance = currentDist;
                    }
                }
            }
        }
        
        return found ? minDistance : -1;
    }
};

File: 3740-minimum-distance-between-three-equal-elements-i/solution.py
from collections import defaultdict
from typing import List

class Solution:
    def minimumDistance(self, nums: List[int]) -> int:
        """
        Calculates the minimum distance of a good tuple.
        Formula simplification: abs(i-j) + abs(j-k) + abs(k-i) = 2 * (max(i,j,k) - min(i,j,k))
        """
        # Dictionary to store a list of indices for each number
        pos_map = defaultdict(list)
        for idx, val in enumerate(nums):
            pos_map[val].append(idx)
            
        min_dist = float('inf')
        
        for val in pos_map:
            indices = pos_map[val]
            # We need at least 3 indices to form a good tuple
            if len(indices) >= 3:
                # To minimize the span, check triplets of consecutive occurrences
                for i in range(len(indices) - 2):
                    # distance = 2 * (last_index - first_index)
                    dist = 2 * (indices[i + 2] - indices[i])
                    if dist < min_dist:
                        min_dist = dist
                        
        return int(min_dist) if min_dist != float('inf') else -1